theseus/optimizer/nonlinear/nonlinear_optimizer.py

```diff
@@ -518,13 +513,43 @@ def _step(
             ignore_mask=converged_indices,
             force_update=force_update,
         )
-        tensor_map = {v.name: v.tensor for v in self._tmp_optim_vars}
+        tensor_dict = {v.name: v.tensor for v in self._tmp_optim_vars}
         with torch.no_grad():
-            err = self.objective.error_squared_norm(tensor_map, also_update=False)
+            err = self.objective.error_squared_norm(tensor_dict, also_update=False) / 2
+        return tensor_dict, err
 
+    # Given descent directions and step sizes, updates the optimization
+    # variables.
+    # Batch indices indicated by `converged_indices` mask are ignored
+    # unless `force_update = True`.
+    # Returns the total error tensor after the update
+    def _step(
+        self,
+        delta: torch.Tensor,
+        previous_err: torch.Tensor,
+        converged_indices: torch.Tensor,
+        force_update: bool,
+        **kwargs,
+    ) -> torch.Tensor:
+        tensor_dict, err = self._compute_retracted_tensors_and_error(
+            delta, converged_indices, force_update
+        )
         reject_indices = self._complete_step(delta, err, previous_err, **kwargs)
-        self.objective.update(tensor_map, batch_ignore_mask=reject_indices)
 
+        if reject_indices is not None and reject_indices.all():
+            # If all steps are marked as rejected, optimizer will think convergence
+            # has happened and stop. Taking a small step prevents this issue
+            tensor_dict, err = self._compute_retracted_tensors_and_error(
```

Comment:
Not sure if the logic is reasonable. According to my experience, a small step might not be sufficient for the optimizer to jump out of the bad trap.

Reply:
I was afraid that an all reject would somehow break the compute graph, and I also wanted to keep things very simple. But I realized that there is an easy way to do this and it can be implemented w/o affecting backward pass. I implemented it in a8d3dc74158097e93c6783a0cb94e78017cc6467. Let me know if this looks OK.

Reply:
I also checked manually that the current end-to-end test in `test_theseus_layer` hits the all reject case, so backward pass does to work with this logic.

Reply:
I mean this b80977dc72c3d2cd45efd31c35760a989cd96b4a